dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/datafile/FileServiceIT.java

```diff
@@ -0,0 +1,213 @@
+                getLatestVersionData(updatedDataset), equalTo(versionDataBefore));
+        assertThat("All files should have been deleted from dataset file list",
+                updatedDataset.getFiles(), empty());
+    }
+
+    @Test
+    public void deleteFiles_forReleasedDataset() {
+        // given
+        publishDataset(DRAFT_DATASET_WITH_FILES_ID);
+        Dataset dataset = datasetDao.find(DRAFT_DATASET_WITH_FILES_ID);
+
+        Tuple2<VersionState, Long> versionDataBefore = getLatestVersionData(dataset);
+
+        DatasetVersion latestVersionBefore = dataset.getLatestVersion();
+        List<FileMetadata> filesToDeleteMetadata = latestVersionBefore.getFileMetadatas();
+
+        // when
+        Set<Dataset> results = fileService.deleteFiles(filesToDeleteMetadata);
+
+        // then
+        Dataset updatedDataset = findNewestResult(results);
+        DatasetVersion versionAfter = updatedDataset.getLatestVersion();
+        Tuple2<VersionState, Long> versionDataAfter = getLatestVersionData(updatedDataset);
+
+        assertThat("State before and after delete should be different and state after delete should be DRAFT",
+                versionDataAfter._1, allOf(not(equalTo(versionDataBefore._1)), equalTo(VersionState.DRAFT)));
+        assertThat("All files metadata should have been deleted from latest version metadata",
+                versionAfter.getFileMetadatas(), empty());
+        assertThat(
+                "Deleted file should be present in dataset file collection, as they're used by previously published version(s)",
+                filesToDeleteMetadata, everyItem(in(extractFileListMetadata(updatedDataset)))
+        );
+    }
+
+    // -------------------- PRIVATE --------------------
+
+    private Tuple2<VersionState, Long> getLatestVersionData(Dataset dataset) {
+        DatasetVersion latestVersion = dataset.getLatestVersion();
+        return Tuple.of(latestVersion.getVersionState(), latestVersion.getVersionNumber());
+    }
+
+    private Dataset findNewestResult(Collection<Dataset> datasets) {
+        return datasets.stream()
+                .reduce((prev, next) -> next.getModificationTime().after(prev.getModificationTime()) ? next : prev)
+                .orElseThrow(() -> new IllegalArgumentException("Cannot find newest result: possibly wrong argument"));
+    }
+
+    private List<FileMetadata> extractFileListMetadata(Dataset dataset) {
+        List<DataFile> files = dataset.getFiles();
+        return files.stream()
+                .map(DataFile::getFileMetadata)
+                .collect(toList());
+    }
+
+    private void publishDataset(Long datasetId) {
+        Dataset dataset = datasetDao.find(datasetId);
+        Dataverse dataverseForDataset = dataset.getDataverseContext();
+        publishDataverse(dataverseForDataset);
+        commandEngine.submit(new PublishDatasetCommand(dataset, requestService.getDataverseRequest(), false));
```

Comment:
This is tight coupling of functionalities. We depend here on a lot to happen correctly. We should try to avoid that.
What I mean is that when publishing command will work stop working we will have chain reaction. Something that test publishing will fail and this service will fail also. In result we would have to spend more time on detecting what actually gone wrong. Not to mention if we would try to execute this with unprivileged user it would also fail.
I propose to be as close to the db operation as possible with that kind of data preparation. That is:
```
dataset.getLatestVersion().setVersionState(RELEASED)
datasetDao.merge(dataset) // currently is not really only a dao (it have some methods with business logic) but in the long run it will be (after a lot of refactor)
```